publisher.go

```diff
@@ -0,0 +1,150 @@
+package stacker
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/anuvu/stacker/lib"
+)
+
+type PublishArgs struct {
+	Config     StackerConfig
+	Debug      bool
+	ShowOnly   bool
+	Substitute []string
+	Tags       []string
+	Url        string
+	Username   string
+	Password   string
+}
+
+// Publisher is responsible for publishing the layers based on stackerfiles
+type Publisher struct {
+	stackerfiles StackerFiles // Keep track of all the Stackerfiles to publish
+	opts         *PublishArgs // Publish options
+}
+
+// NewPublisher initializes a new Publisher struct
+func NewPublisher(opts *PublishArgs) *Publisher {
+	return &Publisher{
+		stackerfiles: make(map[string]*Stackerfile, 1),
+		opts:         opts,
+	}
+}
+
+// Publish layers in a single stackerfile
+func (p *Publisher) Publish(file string) error {
+	opts := p.opts
+
+	// Read stackerfile
+	// Do not call NewStackerfile directly as there may be substitution errors
+	// substitute with the value 'dummy' in case value is not provided by the user
+	sf, err := p.readStackerFile(file, opts.Substitute)
+	if err != nil {
+		return err
+	}
+
+	// Determine list of tags to be used
+	tags := make([]string, len(opts.Tags))
+	copy(tags, opts.Tags)
+
+	// Attempt to produce a git commit tag
+	if ct, err := NewGitLayerTag(sf.referenceDirectory); err == nil {
+		// Add git tag to the list of tags to be used
+		tags = append(tags, ct)
+	}
+
+	if len(tags) == 0 {
+		fmt.Printf("can't save OCI images in %s since list of tags is empty\n", file)
+	}
+
+	// Need to determine if URL is docker/oci or something else
+	is, err := NewImageSource(opts.Url)
+	if err != nil {
+		return err
+	}
+
+	// Iterate through all layers defined in this stackerfile
+	for _, name := range sf.fileOrder {
```

Comment:
What do you think about checking the cache before we push? Just in case something has changed? Or should we not do that?

Reply:
Added a check. By default we verify cache, but I also added a '--force' command line option for override this.